webgpu.h

```diff
@@ -0,0 +1,1025 @@
+    WGPU_VERTEX_FORMAT_USHORT2 = 0x00000008,
+    WGPU_VERTEX_FORMAT_USHORT4 = 0x00000009,
+    WGPU_VERTEX_FORMAT_SHORT2 = 0x0000000A,
+    WGPU_VERTEX_FORMAT_SHORT4 = 0x0000000B,
+    WGPU_VERTEX_FORMAT_USHORT2_NORM = 0x0000000C,
+    WGPU_VERTEX_FORMAT_USHORT4_NORM = 0x0000000D,
+    WGPU_VERTEX_FORMAT_SHORT2_NORM = 0x0000000E,
+    WGPU_VERTEX_FORMAT_SHORT4_NORM = 0x0000000F,
+    WGPU_VERTEX_FORMAT_HALF2 = 0x00000010,
+    WGPU_VERTEX_FORMAT_HALF4 = 0x00000011,
+    WGPU_VERTEX_FORMAT_FLOAT = 0x00000012,
+    WGPU_VERTEX_FORMAT_FLOAT2 = 0x00000013,
+    WGPU_VERTEX_FORMAT_FLOAT3 = 0x00000014,
+    WGPU_VERTEX_FORMAT_FLOAT4 = 0x00000015,
+    WGPU_VERTEX_FORMAT_UINT = 0x00000016,
+    WGPU_VERTEX_FORMAT_UINT2 = 0x00000017,
+    WGPU_VERTEX_FORMAT_UINT3 = 0x00000018,
+    WGPU_VERTEX_FORMAT_UINT4 = 0x00000019,
+    WGPU_VERTEX_FORMAT_INT = 0x0000001A,
+    WGPU_VERTEX_FORMAT_INT2 = 0x0000001B,
+    WGPU_VERTEX_FORMAT_INT3 = 0x0000001C,
+    WGPU_VERTEX_FORMAT_INT4 = 0x0000001D,
+    WGPU_VERTEX_FORMAT_FORCE32 = 0x7FFFFFFF
+} wgpuVertexFormat;
+
+typedef enum {
+    WGPU_BUFFER_USAGE_BIT_NONE = 0x00000000,
+    WGPU_BUFFER_USAGE_BIT_MAP_READ = 0x00000001,
+    WGPU_BUFFER_USAGE_BIT_MAP_WRITE = 0x00000002,
+    WGPU_BUFFER_USAGE_BIT_COPY_SRC = 0x00000004,
+    WGPU_BUFFER_USAGE_BIT_COPY_DST = 0x00000008,
+    WGPU_BUFFER_USAGE_BIT_INDEX = 0x00000010,
+    WGPU_BUFFER_USAGE_BIT_VERTEX = 0x00000020,
+    WGPU_BUFFER_USAGE_BIT_UNIFORM = 0x00000040,
+    WGPU_BUFFER_USAGE_BIT_STORAGE = 0x00000080,
+    WGPU_BUFFER_USAGE_BIT_INDIRECT = 0x00000100,
+    WGPU_BUFFER_USAGE_BIT_FORCE32 = 0x7FFFFFFF
+} wgpuBufferUsageBit;
+
+typedef enum {
+    WGPU_COLOR_WRITE_MASK_NONE = 0x00000000,
+    WGPU_COLOR_WRITE_MASK_RED = 0x00000001,
+    WGPU_COLOR_WRITE_MASK_GREEN = 0x00000002,
+    WGPU_COLOR_WRITE_MASK_BLUE = 0x00000004,
+    WGPU_COLOR_WRITE_MASK_ALPHA = 0x00000008,
+    WGPU_COLOR_WRITE_MASK_ALL = 0x0000000F,
+    WGPU_COLOR_WRITE_MASK_FORCE32 = 0x7FFFFFFF
+} wgpuColorWriteMask;
+
+typedef enum {
+    WGPU_FACE_NONE = 0x00000000,
+    WGPU_FACE_BACK = 0x00000001,
+    WGPU_FACE_FRONT = 0x00000002,
+    WGPU_FACE_BOTH = 0x00000003,
+    WGPU_FACE_FORCE32 = 0x7FFFFFFF
+} wgpuFace;
+
+typedef enum {
+    WGPU_SHADER_STAGE_BIT_NONE = 0x00000000,
+    WGPU_SHADER_STAGE_BIT_VERTEX = 0x00000001,
+    WGPU_SHADER_STAGE_BIT_FRAGMENT = 0x00000002,
+    WGPU_SHADER_STAGE_BIT_COMPUTE = 0x00000004,
+    WGPU_SHADER_STAGE_BIT_FORCE32 = 0x7FFFFFFF
+} wgpuShaderStageBit;
+
+typedef enum {
+    WGPU_TEXTURE_USAGE_BIT_NONE = 0x00000000,
+    WGPU_TEXTURE_USAGE_BIT_COPY_SRC = 0x00000001,
+    WGPU_TEXTURE_USAGE_BIT_COPY_DST = 0x00000002,
+    WGPU_TEXTURE_USAGE_BIT_SAMPLED = 0x00000004,
+    WGPU_TEXTURE_USAGE_BIT_STORAGE = 0x00000008,
+    WGPU_TEXTURE_USAGE_BIT_OUTPUT_ATTACHMENT = 0x00000010,
+    WGPU_TEXTURE_USAGE_BIT_FORCE32 = 0x7FFFFFFF
+} wgpuTextureUsageBit;
+
+
+typedef struct wgpuBindGroupBinding {
+    uint32_t binding;
+    wgpuBuffer buffer;
```

Comment:
`GPUBindingResource` is a sum type that doesn't exist in C/C++ (unions really don't work well for C++ and it implementing a wire with a union type would require doing BGL validation on the wire). Here all the various pieces of `GPUBindingResource` are inlined with exactly one of `buffer`, `sampler` and `textureView` being non-null to pass validation.

Reply:
But C++ does have unions, and all we need is just to tag them.
Not sure what "BGL" is about for validation on the wire, but it sounds like an implementation detail?

Reply:
BGL is shorthand for Bind Group Layout.

Imagine we have the following:

```cpp
struct WGPUBufferBinding {
    WGPUBuffer buffer;
    uint64_t offset;
    uint64_t size;
};

union WGPUBindingResource {
    WGPUBufferBinding buffer;
    WGPUSampler sampler;
    WGPUTextureView textureView;
};

struct WGPUBingGroupBinding {
    uint32_t binding;
    WGPUBindingResource resource;
};
```

In a remoting implementation, the wire format is likely using number handles in different namespaces to serialize `WGPUSampler`, `WGPUBuffer` and `WGPUTextureView` so it can't just memcpy the `WGPUBindGroupBinding` on the wire and it needs to know what member of the union is the relevant one. So there are three choices I can see:

 1. The client of the remoting implementation keeps track of the `WGPUBindGroupLayout` so it can match `WGPUBindGroupBindings` with the corresponding entry and serialize the correct member of the union. This is really sad given how much the API is remotable without keeping state.
 2. A tag is added next to the union that describes which union member to use but this is not as nice, see below.
 3. The union fields are all inlined in `WGPUBindGroupBinding` with only one of the resources being allowed to be non-null (this is checked on the server side).

I'd strongly like to avoid 1) because it is too implicit (it risks crashes when the application uses the wrong resource type) and adds client-side state tracking. And I'd rather have 3) than 2) because the client code is much nicer:

```c
/* Option 2) */
WGPUBindGroupBinding binding = {
    .binding = 0,
    .bindingType = WGPUBindGroupBindingUnionType_Buffer
    .resource = {
        .buffer = {
            .buffer = myBuffer,
            .offset = myOffset,
            .size = mySize,
        }
    }
};

/* Option 3) */
WGPUBindGroupBinding binding = {
    .binding = 0,
    .buffer = myBuffer,
    .offset = myOffset,
    .size = mySize
};
```

Reply:
I was talking about option (2), but thank you for clearly describing this!
It looks like both unions and inlined representations are ugly, but the latter is at least more straightforward. Let's proceed with (3) then, we can work with it.